code/game/machinery/guestpass.dm

```diff
@@ -39,154 +39,182 @@
-//				to_chat(usr, "Printing the log, standby...")
-				//sleep(50)
-				var/obj/item/paper/P = new/obj/item/paper( loc )
-				playsound(loc, 'sound/goonstation/machines/printer_dotmatrix.ogg', 50, 1)
-				P.name = "activity log"
-				P.info = dat
-
-			if("issue")
-				if(giver)
-					var/number = add_zero("[rand(0,9999)]", 4)
-					var/entry = "\[[station_time()]\] Pass #[number] issued by [giver.registered_name] ([giver.assignment]) to [giv_name]. Reason: [reason]. Grants access to following areas: "
-					for(var/i=1 to accesses.len)
-						var/A = accesses[i]
-						if(A)
-							var/area = get_access_desc(A)
-							entry += "[i > 1 ? ", [area]" : "[area]"]"
-					entry += ". Expires at [station_time(world.time + duration*10*60)]."
-					internal_log.Add(entry)
-
-					var/obj/item/card/id/guest/pass = new(src.loc)
-					pass.temp_access = accesses.Copy()
-					pass.registered_name = giv_name
-					pass.expiration_time = world.time + duration*10*60
-					pass.reason = reason
-					pass.name = "guest pass #[number]"
+					scan.forceMove(get_turf(src))
+					scan = null
+				accesses.Cut()
+			else
+				var/obj/item/I = usr.get_active_hand()
+				if(istype(I, /obj/item/card/id))
+					usr.drop_item()
+					I.forceMove(src)
+					scan = I
+		if("mode")
+			mode = !mode
+	if(!scan || !scan.access)
+		return // everything below here requires card auth
+	switch(action)
+		if("giv_name")
+			var/nam = strip_html_simple(input("Person pass is issued to", "Name", giv_name) as text | null)
+			if(nam)
+				giv_name = nam
+		if("reason")
+			var/reas = strip_html_simple(input("Reason why pass is issued", "Reason", reason) as text | null)
+			if(reas)
+				reason = reas
+		if("duration")
+			var/dur = input("Duration (in minutes) during which pass is valid (up to 30 minutes).", "Duration") as num | null
+			if(dur)
+				if(dur > 0 && dur <= 30)
+					duration = dur
 				else
-					to_chat(usr, "<span class='warning'>Cannot issue pass without issuing ID.</span>")
-	updateUsrDialog()
-	return
+					to_chat(usr, "<span class='warning'>Invalid duration.</span>")
+		if("print")
+			var/dat = "<h3>Activity log of guest pass terminal #[uid]</h3><br>"
+			for(var/entry in internal_log)
+				dat += "[entry]<br><hr>"
+			var/obj/item/paper/P = new/obj/item/paper(loc)
+			playsound(loc, 'sound/goonstation/machines/printer_dotmatrix.ogg', 50, TRUE)
+			P.name = "activity log"
+			P.info = dat
+		if("issue")
+			if(!length(accesses))
+				return
+			if(print_cooldown > world.time)
+				return
+			var/number = add_zero("[rand(0, 9999)]", 4)
+			var/entry = "\[[station_time()]\] Pass #[number] issued by [scan.registered_name] ([scan.assignment]) to [giv_name]. Reason: [reason]. Grants access to following areas: "
+			for(var/i in 1 to length(accesses))
+				var/A = accesses[i]
+				if(A)
+					var/area = get_access_desc(A)
+					entry += "[i > 1 ? ", [area]" : "[area]"]"
+			var/obj/item/card/id/guest/pass = new(get_turf(src))
+			pass.temp_access = accesses.Copy()
```

Comment:
Needs to filter out the accesses `scan` may not have, lest it will grant more than technically allowed.

Reply:
That is is checked when they're added to accesses, not here.

Reply:
What happens if you make an AA guest pass with the spare, take it out without touching the selected accesses then issue another guest pass with an assistant's ID?

Reply:
"scan = null" only exists on 121 and 124. In both of those cases it is followed by "accesses.Cut()".